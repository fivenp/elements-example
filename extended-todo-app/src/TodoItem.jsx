import React from 'react'
import PropTypes from 'prop-types'
import Icon from '@allthings/elements/atoms/Icon'
import Text from '@allthings/elements/atoms/Text'
import Checkmark from '@allthings/elements/molecules/Checkmark'
import ListItem from '@allthings/elements/molecules/List/ListItem'
import TextInput from '@allthings/elements/molecules/TextInput'
import { css } from 'glamor'

const { number, string, bool } = PropTypes

const dynamicStyles = (checkmarkNo, icons, iconTransition, slideLeft) => ({
  check: css({
    display: checkmarkNo + ' !important',
  }),
  inline: {
    width: icons,
    transitionDelay: iconTransition,
  },
  remove: {
    right: slideLeft,
  },
  iconBackground: {
    right: slideLeft,
  },
})

const styles = {
  check: css({
    height: '40px',
    width: '40px',
    margin: '3px 0px',
  }),
  inline: css({
    overflow: 'hidden',
    display: 'inline-flex',
  }),
  remove: css({
    float: 'right',
    padding: '5px',
    margin: '3px',
    transition: 'all 300ms ease',
    ':hover': { cursor: 'pointer' },
  }),
  iconBackground: css({
    position: 'absolute',
    padding: '15px',
    float: 'right',
    transition: 'all 300ms ease',
    height: '35.5px',
    width: '40px',
  }),
  list: css({
    display: 'inline-flex',
  }),
  listItem: css({
    minHeight: '50px',
    width: '100%',
  }),
  task: css({
    display: 'inline-flex',
    width: '90%',
    padding: '10px 0',
  }),
  text: css({
    margin: '0px 10px',
    width: '100%',
    height: '50.5px',
    display: 'table',
  }),
  removeDiv: css({
    position: 'relative',
    overflow: 'hidden',
    height: '50.5px',
    width: '53.5px',
    padding: '0px 10px 15px 10px',
  }),
  editDiv: css({
    position: 'relative',
    overflow: 'hidden',
    height: '50.5px',
    width: '53.5px',
    padding: '0px 10px 15px 10px',
  }),
  gray: css({
    backgroundColor: 'gray',
    transition: 'all 300ms ease',
  }),
  red: css({
    backgroundColor: 'red',
    transition: 'all 300ms ease',
  }),
  editField: css({
    padding: 0,
    ':focus': { outline: 'none' },
  }),
  staticText: css({
    display: 'table-cell !important',
    verticalAlign: 'middle',
  }),
}

const classNames = {
  edit: ['icon-div', 'edit-icon'],
  iconDiv: ['icon-div', 'main-icon-div'],
  remove: ['icon-div', 'remove-icon'],
}

class TodoItem extends React.Component {
  static propTypes = {
    id: number.isRequired,
    children: string.isRequired,
    done: bool.isRequired,
  }

  state = {
    touchStart: undefined,
    icons: 0,
    iconTransition: '',
    slideLeft: -70,
    checkmark: 'flex',
  }

  handleRemove = () => {
    this.closeIcons('')
    this.props.handleRemove(this.props.id)
  }

  handleEdit = () => {
    this.closeIcons('')
    this.props.onDoubleClick(this.props.id)
  }

  handleEditComplete = event => {
    this.closeIcons('')
    this.props.doneEditting(event, this.props.id)
  }

  handleCheckClick = () => this.props.checkmarkClicked(this.props.id)

  openIcons = () =>
    this.setState({
      checkmark: 'none',
      slideLeft: 0,
      icons: 150,
      iconTransition: '',
    })

  closeIcons = transition =>
    this.setState({
      checkmark: 'flex',
      slideLeft: -70,
      icons: 0,
      iconTransition: transition || '',
      doubleClicked: false,
    })

  handleTouchStart = event => {
    this.setState({
      touchStart: event.changedTouches[0].clientX,
    })
  }

  handleTouchMove = event => {
    const { slideLeft, touchStart } = this.state
    const { onSlideLeft, id } = this.props
    const touchObj = event.changedTouches[0].clientX
    const dist = parseInt(touchObj) - touchStart
    if (dist < -10) {
      this.openIcons()
      onSlideLeft(id, 'open')
    } else if (dist > 5) {
      this.closeIcons('200ms')
      onSlideLeft(id, 'close')
    }
  }

  render() {
    const { id, children, done, doubleClicked, iconOpen } = this.props
    const { icons, iconTransition, slideLeft, checkmark } = this.state
    const checkmarkNo = iconOpen || doubleClicked ? 'none' : checkmark

    const stylesDynamic = !iconOpen
      ? dynamicStyles('flex', 0, '', -70)
      : dynamicStyles(checkmarkNo, icons, iconTransition, slideLeft)

    return (
      <div {...styles.list}>
        <ListItem {...styles.listItem}>
          <div {...styles.task}>
            <Checkmark
              id={`checkmark${id}`}
              checked={done}
              onClick={this.handleCheckClick}
              {...stylesDynamic.check}
              {...styles.check}
            />
            <div
              className="touch"
              onDoubleClick={this.handleEdit}
              onTouchStart={this.handleTouchStart}
              onTouchMove={this.handleTouchMove}
              {...styles.text}
            >
              {doubleClicked ? (
                <TextInput
                  id={id}
                  className="edit-input"
                  name="edit"
                  defaultValue={children}
                  placeholder="Edit todo"
                  onKeyPress={this.handleEditComplete}
                  {...styles.editField}
                />
              ) : (
                <Text autoBreak={true} id={id} {...styles.staticText}>
                  {children}
                </Text>
              )}
            </div>
          </div>
          <div
            className= {classNames.iconDiv.join(' ')}
            style={stylesDynamic.inline}
            {...styles.inline}
          >
            <div className="icon-div" {...styles.editDiv}>
              <div
                className="icon-div"
                style={stylesDynamic.iconBackground}
                {...styles.iconBackground}
                {...styles.gray}
              >
                <Icon
                  className={classNames.edit.join(' ')}
                  name="edit"
                  size="m"
                  color="white"
                  onClick={this.handleEdit}
                  style={stylesDynamic.remove}
                  {...styles.remove}
                />
              </div>
            </div>
            <div className="icon-div" {...styles.removeDiv}>
              <div
                className="icon-div"
                style={stylesDynamic.iconBackground}
                {...styles.iconBackground}
                {...styles.red}
              >
                <Icon
                  className={classNames.remove.join(' ')}
                  name="remove-light-filled"
                  size="m"
                  color="white"
                  onClick={this.handleRemove}
                  style={stylesDynamic.remove}
                  {...styles.remove}
                />
              </div>
            </div>
          </div>
        </ListItem>
      </div>
    )
  }
}

export default TodoItem
