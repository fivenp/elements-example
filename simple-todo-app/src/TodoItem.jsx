import React from 'react'
import PropTypes from 'prop-types'
import { Checkmark, Inset, ListItem, Text } from '@allthings/elements'

class TodoItem extends React.Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    children: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired
  }

  handleClick = () => this.props.onClick(this.props.id)

  render() {
    return (
      <ListItem onClick={this.handleClick}>
        <Checkmark checked={this.props.done} />
        <Inset>
          <Text>{this.props.children}</Text>
        </Inset>
      </ListItem>
    )
  }
}

export default TodoItem
