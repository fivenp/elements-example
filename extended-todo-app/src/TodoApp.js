import React, { Component } from 'react'
import { ColorPalette } from '@allthings/colors'
import Icon from '@allthings/elements/atoms/Icon'
import Inset from '@allthings/elements/atoms/Inset'
import Text from '@allthings/elements/atoms/Text'
import View from '@allthings/elements/atoms/View'
import Button from '@allthings/elements/molecules/Button'
import Card from '@allthings/elements/molecules/Card'
import TextInput from '@allthings/elements/molecules/TextInput'
import GroupedCardList from '@allthings/elements/organisms/CardList/GroupedCardList'
import TitleBar from '@allthings/elements/organisms/TitleBar'
import ResourceProvider from '@allthings/elements/behaviour/ResourceProvider'
import ThemeProvider from '@allthings/elements/behaviour/ThemeProvider'
import TodoItem from './TodoItem'
import { css } from 'glamor'

const styles = {
  create: css({
    ':hover': { cursor: 'pointer' },
  }),
  inset: css({
    width: '93%',
  }),
  new: css({
    padding: 0,
    width: '100%',
    display: 'flex',
  }),
  iconStyle: css({
    display: 'inline-block',
    marginRight: '10px',
    position: 'absolute',
  }),
  titleBar: css({
    padding: '10px 15px 10px 0px',
  }),
  titleText: css({
    display: 'inline-block',
    padding: '0 35px',
    position: 'relative',
  }),
  buttonDiv: css({
    textAlign: 'center',
  }),
  button: css({
    margin: '3px',
  }),
  textInput: css({
    display: 'inline-block',
    width: '100%',
    ':focus': { outline: 'none' },
  }),
  add: css({
    display: 'inline-block',
    float: 'right',
    margin: '15px 10px 15px 0px',
  }),
}

const buttonOpacity = 'rgba(232, 76, 61, 0.5)'

const DemoTheme = {
  primary: ColorPalette.red,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
}

class App extends Component {
  state = {
    todos: [
      {
        id: 0,
        text: 'Search for cocktail recipes for the party',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
      {
        id: 1,
        text: 'Create nice invitation cards',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
      {
        id: 2,
        text: 'Ask some people to bring some finger food',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
    ],
    filter: 'all',
    textDisable: true,
    newText: '',
  }

  updateInput = event => {
    this.setState({ newText: this.textInput.input.value })
  }

  componentDidMount = () => {
    document.addEventListener('touchstart', this.finishEdit)
    document.addEventListener('mousedown', this.finishEdit)
  }

  handleRemove = id => {
    const newTodo = this.state.todos.slice()
    newTodo.splice(id, 1)
    this.setState({ todos: newTodo })
  }

  handleDoubleClick = id => {
    const newTodo = this.state.todos.slice()
    newTodo[id].doubleClicked = true
    this.setState({ todos: newTodo })
  }

  handleSliding = (id, type) => {
    const newTodo = [...this.state.todos]
    newTodo[id].iconOpen = type === 'open' ? true : false
    this.setState({ todos: newTodo })
  }

  checkmarkClicked = id => {
    const newTodo = [...this.state.todos]
    newTodo[id].done = !this.state.todos[id].done
    this.setState({ todos: newTodo })
  }

  handleKeyPress = event => {
    const text = event.target.value
    if (event.key === 'Enter' && text) {
      this.addTodo()
    } else {
      const textExists = this.textInput.input.value
      if (textExists && this.state.textDisable) {
        this.setState({ textDisable: false })
      } else if (!textExists && !this.state.textDisable) {
        this.setState({ textDisable: true })
      }
    }
  }

  addTodo = () => {
    const text = this.textInput.input.value
    this.setState({
      todos: [
        ...this.state.todos,
        {
          text,
          done: false,
          doubleClicked: false,
          checkmark: 'flex',
          iconOpen: false,
        },
      ],
      textDisable: true,
    })
    this.setState({ newText: '' })
    this.textInput.input.blur()
  }

  changeFilter = ({ target: { className } }) => {
    this.setState({ filter: className })
  }

  focusInput = () => {
    this.textInput.input.focus()
  }

  doneEditting = (event, id) => {
    const text = event.target.value
    if (event.key === 'Enter' && text) {
      const newTodo = [...this.state.todos]
      newTodo[id] = {
        ...newTodo[id],
        doubleClicked: false,
        text,
        checkmark: 'flex',
        iconOpen: false,
      }
      this.setState({ todos: newTodo })
    }
  }

  finishEdit = event => {
    const oldTodo = [...this.state.todos]
    const eventSrc = event.srcElement
    const { id: eventId, nodeName: eventNode, className: eventClass } = eventSrc
    const newTodo = oldTodo.map((todo, index) => {
      const { doubleClicked, iconOpen } = todo
      if (eventClass === 'icon-div' || eventNode === ('svg' || 'path')) {
        return todo
      } else if ((doubleClicked || iconOpen) && index.toString() === eventId) {
        return todo
      } else if (doubleClicked && eventClass !== 'icon-div') {
        todo = {
          ...todo,
          doubleClicked: false,
          checkmark: 'flex',
          iconOpen: false,
          text: document.getElementById(index).value,
        }
        return todo
      } else if (iconOpen && eventClass !== 'icon-div') {
        todo = {
          ...todo,
          doubleClicked: false,
          checkmark: 'flex',
          iconOpen: false,
        }
        return todo
      } else {
        return todo
      }
    })
    this.setState({ todos: newTodo })
  }

  render() {
    const { todos, filter, textDisable } = this.state
    const incomplete = todos.reduce(
      (count, todo) => (todo.done ? count : count + 1),
      0
    )
    const incompleteNum =
      incomplete > 0 ? `Incomplete (${incomplete})` : 'Incomplete'

    return (
      <div className="App">
        <ResourceProvider>
          <ThemeProvider theme={DemoTheme}>
            <View>
              <TitleBar {...styles.titleBar}>
                <Inset {...styles.inset}>
                  <Icon
                    name="book-open"
                    size="m"
                    color="white"
                    {...styles.iconStyle}
                  />
                  <Text color="contrast" {...styles.titleText}>
                    My Todos
                  </Text>
                </Inset>
                <Icon
                  name="plus-light"
                  size="m"
                  color="#FFFFFF"
                  onClick={this.focusInput}
                  {...styles.create}
                />
              </TitleBar>
              <Card>
                <div {...styles.new}>
                  <div {...styles.textInput}>
                    <TextInput
                      id="new"
                      name="new"
                      ref={node => (this.textInput = node)}
                      placeholder="Add new"
                      value={this.state.newText}
                      onKeyUp={this.handleKeyPress}
                      onChange={this.updateInput.bind(this)}
                      {...styles.textInput}
                    />
                  </div>
                  <Button
                    id="add-task"
                    onClick={this.addTodo}
                    disabled={textDisable}
                    {...styles.add}
                  >
                    add
                  </Button>
                </div>
                <div {...styles.buttonDiv}>
                  <Button
                    className="all"
                    onClick={this.changeFilter}
                    backgroundColor={filter !== 'all' ? buttonOpacity : ''}
                    {...styles.button}
                  >
                    <span className="all">All</span>
                  </Button>
                  <Button
                    className="incomplete"
                    onClick={this.changeFilter}
                    backgroundColor={
                      filter !== 'incomplete' ? buttonOpacity : ''
                    }
                    {...styles.button}
                  >
                    <span className="incomplete">{incompleteNum}</span>
                  </Button>
                  <Button
                    id="completed-button"
                    className="completed"
                    onClick={this.changeFilter}
                    backgroundColor={
                      filter !== 'completed' ? buttonOpacity : ''
                    }
                    {...styles.button}
                  >
                    <span className="completed">Completed</span>
                  </Button>
                </div>
                <GroupedCardList>
                  {todos.map((todo, index) => {
                    const {
                      done,
                      checkmark,
                      iconOpen,
                      doubleClicked,
                      text,
                    } = todo
                    const thisTodoItem = (
                      <TodoItem
                        handleRemove={this.handleRemove}
                        id={index}
                        done={done}
                        checkmarkDisplay={checkmark}
                        iconOpen={iconOpen}
                        doubleClicked={doubleClicked}
                        key={'checkbox' + index}
                        onDoubleClick={this.handleDoubleClick}
                        onSlideLeft={this.handleSliding}
                        doneEditting={this.doneEditting}
                        checkmarkClicked={this.checkmarkClicked}
                      >
                        {text}
                      </TodoItem>
                    )
                    if (
                      (filter === 'incomplete' && done === false) ||
                      ((filter === 'completed' && done === true) ||
                        filter === 'all')
                    ) {
                      return thisTodoItem
                    }
                  })}
                </GroupedCardList>
              </Card>
            </View>
          </ThemeProvider>
        </ResourceProvider>
      </div>
    )
  }
}

export default App
