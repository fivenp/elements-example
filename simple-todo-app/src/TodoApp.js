import React, { Component } from 'react'
import { ColorPalette } from '@allthings/colors'
import { Card, List, Inset, ResourceProvider, TitleBar, Text, ThemeProvider } from '@allthings/elements'
import TodoItem from './TodoItem'

const DemoTheme = {
  primary: ColorPalette.red,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright
}

class App extends Component {
  state = {
    todos: [
      { id: 1, text: 'Search for cocktail recipes for the party', done: false },
      { id: 2, text: 'Create nice invitation cards', done: false },
      { id: 3, text: 'Ask some people to bring some finger food', done: false }
    ]
  }

  handleClick = id => {
    this.setState(({ todos }) => ({
      todos: todos.map(todo => {
        if (todo.id === id) {
          return {
            ...todo,
            done: !todo.done
          }
        } else {
          return todo
        }
      })
    }))
  }

  render() {
    return (
      <div className="App">
        <ThemeProvider theme={DemoTheme}>
          <ResourceProvider>
            <TitleBar>
              <Inset>
                <Text strong color="white">
                  Todos
                </Text>
              </Inset>
            </TitleBar>
            <Card>
              <List>
                {this.state.todos.map(todo => (
                  <TodoItem
                    onClick={this.handleClick}
                    id={todo.id}
                    done={todo.done}
                  >
                    {todo.text}
                  </TodoItem>
                ))}
              </List>
            </Card>
          </ResourceProvider>
        </ThemeProvider>
      </div>
    )
  }
}

export default App
