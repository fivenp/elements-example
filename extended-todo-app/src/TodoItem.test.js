import React from 'react'
import { shallow, mount, render } from 'enzyme'
import TodoItem from './TodoItem'
import TodoApp from './TodoApp'

describe('TodoItem', () => {
  it('it should match snapshot', () => {
    const wrapper = shallow(
      <TodoItem id={0} done={true}>
        Some Text
      </TodoItem>
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('TodoItem buttons', () => {
  const component = mount(<TodoApp />).setState({
    todos: [
      {
        id: 0,
        text: 'Text 0',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
      {
        id: 1,
        text: 'Text 1',
        done: false,
      },
      {
        id: 2,
        text: 'Text 2',
        done: false,
      },
    ],
    filter: 'all',
    textDisable: true,
    newText: '',
  })

  it('should set done = false if checkmark clicked', () => {
    component
      .find(TodoItem)
      .first()
      .find('#checkmark0')
      .first()
      .simulate('click')
    expect(
      component
        .find(TodoItem)
        .first()
        .prop('done')
    ).toBe(true)
  })

  it('should correctly change width of div containing icons if slided >10px', () => {
    const touchDiv = component
      .find(TodoItem)
      .first()
      .find('div.touch')

    //doesn't do anything if slide is =<10px
    touchDiv
      .simulate('touchStart', {
        changedTouches: [{ clientX: 243, clientY: 234 }],
      })
      .simulate('touchMove', {
        changedTouches: [{ clientX: 233, clientY: 234 }],
      })
    expect(
      component
        .find(TodoItem)
        .first()
        .find('div.main-icon-div')
        .props().style.width
    ).toEqual(0)

    //shows icon div if slided right >10px
    touchDiv
      .simulate('touchStart', {
        changedTouches: [{ clientX: 243, clientY: 234 }],
      })
      .simulate('touchMove', {
        changedTouches: [{ clientX: 230, clientY: 234 }],
      })
    expect(
      component
        .find(TodoItem)
        .first()
        .find('div.main-icon-div')
        .props().style.width
    ).toEqual(150)

    //puts away icon div if slided right >10px
    touchDiv
      .simulate('touchStart', {
        changedTouches: [{ clientX: 230, clientY: 234 }],
      })
      .simulate('touchMove', {
        changedTouches: [{ clientX: 243, clientY: 234 }],
      })
    expect(
      component
        .find(TodoItem)
        .first()
        .find('div.main-icon-div')
        .props().style.width
    ).toEqual(0)
  })

  it('should change text to input if edit is clicked', () => {
    component
      .find(TodoItem)
      .first()
      .find('div.edit-icon')
      .simulate('click')
    expect(component.find('input.edit-input')).toHaveLength(1)
  })

  it('should delete task if remove is clicked', () => {
    component
      .find(TodoItem)
      .first()
      .find('div.remove-icon')
      .simulate('click')
    expect(component.state().todos).toHaveLength(2)
  })
})
