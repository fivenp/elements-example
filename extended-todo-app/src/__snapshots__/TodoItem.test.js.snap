// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem it should match snapshot 1`] = `
<div
  data-css-h0biv2=""
>
  <ListItem
    backgroundColor="white"
    data-css-zh5k6q=""
    padded={true}
  >
    <div
      data-css-154b89o=""
    >
      <_class
        checked={true}
        data-css-13isb4y=""
        data-css-p2i0ep=""
        id="checkmark0"
        onClick={[Function]}
      />
      <div
        className="touch"
        data-css-leaz66=""
        onDoubleClick={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <Text
          autoBreak={true}
          block={true}
          color="text"
          data-css-93utwa=""
          id={0}
          size="l"
        >
          Some Text
        </Text>
      </div>
    </div>
    <div
      className="icon-div main-icon-div"
      data-css-1r5myjm=""
      style={
        Object {
          "transitionDelay": "",
          "width": 0,
        }
      }
    >
      <div
        className="icon-div"
        data-css-b7lh05=""
      >
        <div
          className="icon-div"
          data-css-1nax9fv=""
          data-css-pmoomt=""
          style={
            Object {
              "right": -70,
            }
          }
        >
          <Icon
            className="icon-div edit-icon"
            color="white"
            data-css-fvks2t=""
            name="edit"
            onClick={[Function]}
            size="m"
            style={
              Object {
                "right": -70,
              }
            }
          />
        </div>
      </div>
      <div
        className="icon-div"
        data-css-b7lh05=""
      >
        <div
          className="icon-div"
          data-css-1nax9fv=""
          data-css-83ugar=""
          style={
            Object {
              "right": -70,
            }
          }
        >
          <Icon
            className="icon-div remove-icon"
            color="white"
            data-css-fvks2t=""
            name="remove-light-filled"
            onClick={[Function]}
            size="m"
            style={
              Object {
                "right": -70,
              }
            }
          />
        </div>
      </div>
    </div>
  </ListItem>
</div>
`;
